File: src/packages/sdk/pynomaly_sdk/data_science.py
"""
Pynomaly SDK Data Science API

High-level API for data science operations including detector management,
anomaly detection, and experiment tracking.
"""

from typing import Optional, Dict, Any, List, Union
from uuid import UUID
import pandas as pd
import numpy as np

from .models import (
    DetectorConfig,
    Dataset,
    DetectionResult,
    ExperimentConfig,
    ModelMetrics,
    TrainingJob,
    PaginatedResponse
)
from .exceptions import ValidationError, APIError, ResourceNotFoundError


class DataScienceAPI:
    """
    High-level API for data science operations.
    
    Provides convenient methods for common anomaly detection workflows
    including detector management, data processing, and experiment tracking.
    """
    
    def __init__(self, client):
        """Initialize with a PynomalyClient instance."""
        self._client = client
    
    # Detector Management
    
    async def create_detector(
        self,
        name: str,
        config: DetectorConfig,
        description: Optional[str] = None,
        tags: Optional[List[str]] = None
    ) -> Dict[str, Any]:
        """
        Create a new anomaly detector.
        
        Args:
            name: Detector name
            config: Detector configuration
            description: Optional description
            tags: Optional tags for organization
            
        Returns:
            Created detector information
            
        Raises:
            ValidationError: If configuration is invalid
            APIError: If creation fails
        """
        payload = {
            "name": name,
            "algorithm_name": config.algorithm_name,
            "hyperparameters": config.hyperparameters,
            "contamination_rate": config.contamination_rate,
            "random_state": config.random_state,
            "n_jobs": config.n_jobs
        }
        
        if description:
            payload["description"] = description
        if tags:
            payload["tags"] = tags
            
        response = await self._client._request("POST", "/detectors", json_data=payload)
        return response.data
    
    async def list_detectors(
        self,
        page: int = 1,
        page_size: int = 20,
        algorithm_name: Optional[str] = None,
        tags: Optional[List[str]] = None
    ) -> PaginatedResponse:
        """
        List available detectors.
        
        Args:
            page: Page number (1-based)
            page_size: Number of detectors per page
            algorithm_name: Filter by algorithm name
            tags: Filter by tags
            
        Returns:
            Paginated list of detectors
        """
        params = {
            "page": page,
            "page_size": page_size
        }
        
        if algorithm_name:
            params["algorithm_name"] = algorithm_name
        if tags:
            params["tags"] = ",".join(tags)
            
        response = await self._client._request("GET", "/detectors", params=params)
        return PaginatedResponse(**response.data)
    
    async def get_detector(self, detector_id: Union[str, UUID]) -> Dict[str, Any]:
        """
        Get detector details by ID.
        
        Args:
            detector_id: Detector ID
            
        Returns:
            Detector information
            
        Raises:
            ResourceNotFoundError: If detector not found
        """
        try:
            response = await self._client._request("GET", f"/detectors/{detector_id}")
            return response.data
        except APIError as e:
            if e.status_code == 404:
                raise ResourceNotFoundError("Detector", str(detector_id))
            raise
    
    async def update_detector(
        self,
        detector_id: Union[str, UUID],
        name: Optional[str] = None,
        config: Optional[DetectorConfig] = None,
        description: Optional[str] = None,
        tags: Optional[List[str]] = None
    ) -> Dict[str, Any]:
        """
        Update detector configuration.
        
        Args:
            detector_id: Detector ID
            name: New name (optional)
            config: New configuration (optional)
            description: New description (optional)
            tags: New tags (optional)
            
        Returns:
            Updated detector information
        """
        payload = {}
        
        if name is not None:
            payload["name"] = name
        if config is not None:
            payload.update({
                "algorithm_name": config.algorithm_name,
                "hyperparameters": config.hyperparameters,
                "contamination_rate": config.contamination_rate,
                "random_state": config.random_state,
                "n_jobs": config.n_jobs
            })
        if description is not None:
            payload["description"] = description
        if tags is not None:
            payload["tags"] = tags
            
        response = await self._client._request("PUT", f"/detectors/{detector_id}", json_data=payload)
        return response.data
    
    async def delete_detector(self, detector_id: Union[str, UUID]) -> bool:
        """
        Delete a detector.
        
        Args:
            detector_id: Detector ID
            
        Returns:
            True if deleted successfully
        """
        try:
            await self._client._request("DELETE", f"/detectors/{detector_id}")
            return True
        except APIError as e:
            if e.status_code == 404:
                raise ResourceNotFoundError("Detector", str(detector_id))
            raise
    
    # Training Operations
    
    async def train_detector(
        self,
        detector_id: Union[str, UUID],
        dataset: Union[Dataset, pd.DataFrame, np.ndarray],
        job_name: Optional[str] = None
    ) -> TrainingJob:
        """
        Train a detector with provided data.
        
        Args:
            detector_id: Detector ID to train
            dataset: Training dataset
            job_name: Optional job name
            
        Returns:
            Training job information
        """
        # Convert dataset to API format
        if isinstance(dataset, (pd.DataFrame, np.ndarray)):
            dataset = Dataset.from_dataframe("training_data", dataset) if isinstance(dataset, pd.DataFrame) else Dataset.from_numpy("training_data", dataset)
        
        payload = {
            "detector_id": str(detector_id),
            "dataset": dataset.to_dict()
        }
        
        if job_name:
            payload["job_name"] = job_name
            
        response = await self._client._request("POST", "/training/jobs", json_data=payload)
        return TrainingJob(**response.data)
    
    async def get_training_job(self, job_id: Union[str, UUID]) -> TrainingJob:
        """
        Get training job status and results.
        
        Args:
            job_id: Training job ID
            
        Returns:
            Training job information
        """
        try:
            response = await self._client._request("GET", f"/training/jobs/{job_id}")
            return TrainingJob(**response.data)
        except APIError as e:
            if e.status_code == 404:
                raise ResourceNotFoundError("Training Job", str(job_id))
            raise
    
    async def list_training_jobs(
        self,
        detector_id: Optional[Union[str, UUID]] = None,
        status: Optional[str] = None,
        page: int = 1,
        page_size: int = 20
    ) -> PaginatedResponse:
        """
        List training jobs with optional filtering.
        
        Args:
            detector_id: Filter by detector ID
            status: Filter by job status
            page: Page number
            page_size: Jobs per page
            
        Returns:
            Paginated list of training jobs
        """
        params = {
            "page": page,
            "page_size": page_size
        }
        
        if detector_id:
            params["detector_id"] = str(detector_id)
        if status:
            params["status"] = status
            
        response = await self._client._request("GET", "/training/jobs", params=params)
        return PaginatedResponse(**response.data)
    
    # Detection Operations
    
    async def detect_anomalies(
        self,
        detector_id: Union[str, UUID],
        data: Union[Dataset, pd.DataFrame, np.ndarray],
        return_scores: bool = True,
        threshold: Optional[float] = None
    ) -> DetectionResult:
        """
        Detect anomalies in data using a trained detector.
        
        Args:
            detector_id: ID of trained detector
            data: Data to analyze
            return_scores: Whether to return anomaly scores
            threshold: Custom decision threshold
            
        Returns:
            Detection results
            
        Raises:
            ValidationError: If data format is invalid
            ResourceNotFoundError: If detector not found
        """
        # Convert data to API format
        if isinstance(data, (pd.DataFrame, np.ndarray)):
            data = Dataset.from_dataframe("detection_data", data) if isinstance(data, pd.DataFrame) else Dataset.from_numpy("detection_data", data)
        
        payload = {
            "detector_id": str(detector_id),
            "dataset": data.to_dict(),
            "return_scores": return_scores
        }
        
        if threshold is not None:
            payload["threshold"] = threshold
            
        try:
            response = await self._client._request("POST", "/detection/predict", json_data=payload)
            return DetectionResult(**response.data)
        except APIError as e:
            if e.status_code == 404:
                raise ResourceNotFoundError("Detector", str(detector_id))
            raise
    
    async def batch_detect(
        self,
        detector_id: Union[str, UUID],
        datasets: List[Union[Dataset, pd.DataFrame, np.ndarray]],
        job_name: Optional[str] = None
    ) -> Dict[str, Any]:
        """
        Submit batch detection job for multiple datasets.
        
        Args:
            detector_id: ID of trained detector
            datasets: List of datasets to process
            job_name: Optional job name
            
        Returns:
            Batch job information
        """
        # Convert datasets to API format
        converted_datasets = []
        for i, dataset in enumerate(datasets):
            if isinstance(dataset, (pd.DataFrame, np.ndarray)):
                dataset = Dataset.from_dataframe(f"batch_data_{i}", dataset) if isinstance(dataset, pd.DataFrame) else Dataset.from_numpy(f"batch_data_{i}", dataset)
            converted_datasets.append(dataset.to_dict())
        
        payload = {
            "detector_id": str(detector_id),
            "datasets": converted_datasets
        }
        
        if job_name:
            payload["job_name"] = job_name
            
        response = await self._client._request("POST", "/detection/batch", json_data=payload)
        return response.data
    
    # Experiment Management
    
    async def create_experiment(
        self,
        config: ExperimentConfig,
        dataset: Union[Dataset, pd.DataFrame, np.ndarray]
    ) -> Dict[str, Any]:
        """
        Create and run a new experiment.
        
        Args:
            config: Experiment configuration
            dataset: Dataset for the experiment
            
        Returns:
            Experiment information
        """
        # Convert dataset to API format
        if isinstance(dataset, (pd.DataFrame, np.ndarray)):
            dataset = Dataset.from_dataframe("experiment_data", dataset) if isinstance(dataset, pd.DataFrame) else Dataset.from_numpy("experiment_data", dataset)
        
        payload = {
            "name": config.name,
            "description": config.description,
            "algorithm_configs": [conf.dict() for conf in config.algorithm_configs],
            "evaluation_metrics": config.evaluation_metrics,
            "cross_validation_folds": config.cross_validation_folds,
            "random_state": config.random_state,
            "parallel_jobs": config.parallel_jobs,
            "optimization_enabled": config.optimization_enabled,
            "optimization_trials": config.optimization_trials,
            "optimization_timeout": config.optimization_timeout,
            "dataset": dataset.to_dict()
        }
        
        response = await self._client._request("POST", "/experiments", json_data=payload)
        return response.data
    
    async def get_experiment(self, experiment_id: Union[str, UUID]) -> Dict[str, Any]:
        """
        Get experiment details and results.
        
        Args:
            experiment_id: Experiment ID
            
        Returns:
            Experiment information
        """
        try:
            response = await self._client._request("GET", f"/experiments/{experiment_id}")
            return response.data
        except APIError as e:
            if e.status_code == 404:
                raise ResourceNotFoundError("Experiment", str(experiment_id))
            raise
    
    async def list_experiments(
        self,
        page: int = 1,
        page_size: int = 20,
        status: Optional[str] = None
    ) -> PaginatedResponse:
        """
        List experiments with optional filtering.
        
        Args:
            page: Page number
            page_size: Experiments per page
            status: Filter by status
            
        Returns:
            Paginated list of experiments
        """
        params = {
            "page": page,
            "page_size": page_size
        }
        
        if status:
            params["status"] = status
            
        response = await self._client._request("GET", "/experiments", params=params)
        return PaginatedResponse(**response.data)
    
    # Utility Methods
    
    async def validate_dataset(
        self,
        dataset: Union[Dataset, pd.DataFrame, np.ndarray],
        algorithm_name: Optional[str] = None
    ) -> Dict[str, Any]:
        """
        Validate dataset format and compatibility.
        
        Args:
            dataset: Dataset to validate
            algorithm_name: Optional algorithm to check compatibility
            
        Returns:
            Validation results
        """
        # Convert dataset to API format
        if isinstance(dataset, (pd.DataFrame, np.ndarray)):
            dataset = Dataset.from_dataframe("validation_data", dataset) if isinstance(dataset, pd.DataFrame) else Dataset.from_numpy("validation_data", dataset)
        
        payload = {
            "dataset": dataset.to_dict()
        }
        
        if algorithm_name:
            payload["algorithm_name"] = algorithm_name
            
        response = await self._client._request("POST", "/validation/dataset", json_data=payload)
        return response.data
    
    async def get_supported_algorithms(self) -> List[Dict[str, Any]]:
        """
        Get list of supported algorithms and their configurations.
        
        Returns:
            List of algorithm information
        """
        response = await self._client._request("GET", "/algorithms")
        return response.data.get("algorithms", [])
    
    async def get_algorithm_info(self, algorithm_name: str) -> Dict[str, Any]:
        """
        Get detailed information about a specific algorithm.
        
        Args:
            algorithm_name: Name of the algorithm
            
        Returns:
            Algorithm details and parameter information
        """
        try:
            response = await self._client._request("GET", f"/algorithms/{algorithm_name}")
            return response.data
        except APIError as e:
            if e.status_code == 404:
                raise ResourceNotFoundError("Algorithm", algorithm_name)
            raise

File: src/packages/sdk/pynomaly-js-sdk/src/core/dataScience.ts
/**
 * Pynomaly JavaScript SDK Data Science API
 * 
 * High-level API for data science operations including detector management,
 * anomaly detection, and experiment tracking for web applications.
 */

import {
  DetectorConfig,
  Dataset,
  DetectionResult,
  ExperimentConfig,
  ModelMetrics,
  TrainingJob,
  PaginatedResponse,
  ApiResponse
} from '../types';
import { ValidationError, APIError, ResourceNotFoundError } from './errors';
import type { PynomalyClient } from './client';

/**
 * High-level API for data science operations in web environments.
 * 
 * Provides convenient methods for common anomaly detection workflows
 * including detector management, data processing, and experiment tracking.
 */
export class DataScienceAPI {
  constructor(private client: PynomalyClient) {}

  // Detector Management

  /**
   * Create a new anomaly detector.
   * 
   * @param name Detector name
   * @param config Detector configuration
   * @param description Optional description
   * @param tags Optional tags for organization
   * @returns Promise resolving to created detector information
   */
  async createDetector(
    name: string,
    config: DetectorConfig,
    description?: string,
    tags?: string[]
  ): Promise<any> {
    const payload = {
      name,
      algorithm_name: config.algorithmName,
      hyperparameters: config.hyperparameters || {},
      contamination_rate: config.contaminationRate || 0.1,
      random_state: config.randomState,
      n_jobs: config.nJobs || 1,
      ...(description && { description }),
      ...(tags && { tags })
    };

    const response = await this.client.request({
      method: 'POST',
      url: '/detectors',
      data: payload
    });

    return response.data;
  }

  /**
   * List available detectors with pagination and filtering.
   * 
   * @param options List options
   * @returns Promise resolving to paginated list of detectors
   */
  async listDetectors(options: {
    page?: number;
    pageSize?: number;
    algorithmName?: string;
    tags?: string[];
  } = {}): Promise<PaginatedResponse> {
    const params: Record<string, any> = {
      page: options.page || 1,
      page_size: options.pageSize || 20
    };

    if (options.algorithmName) {
      params.algorithm_name = options.algorithmName;
    }
    if (options.tags) {
      params.tags = options.tags.join(',');
    }

    const response = await this.client.request({
      method: 'GET',
      url: '/detectors',
      params
    });

    return response.data;
  }

  /**
   * Get detector details by ID.
   * 
   * @param detectorId Detector ID
   * @returns Promise resolving to detector information
   */
  async getDetector(detectorId: string): Promise<any> {
    try {
      const response = await this.client.request({
        method: 'GET',
        url: `/detectors/${detectorId}`
      });
      return response.data;
    } catch (error: any) {
      if (error.statusCode === 404) {
        throw new ResourceNotFoundError('Detector', detectorId);
      }
      throw error;
    }
  }

  /**
   * Update detector configuration.
   * 
   * @param detectorId Detector ID
   * @param updates Update options
   * @returns Promise resolving to updated detector information
   */
  async updateDetector(
    detectorId: string,
    updates: {
      name?: string;
      config?: DetectorConfig;
      description?: string;
      tags?: string[];
    }
  ): Promise<any> {
    const payload: Record<string, any> = {};

    if (updates.name) payload.name = updates.name;
    if (updates.config) {
      payload.algorithm_name = updates.config.algorithmName;
      payload.hyperparameters = updates.config.hyperparameters;
      payload.contamination_rate = updates.config.contaminationRate;
      payload.random_state = updates.config.randomState;
      payload.n_jobs = updates.config.nJobs;
    }
    if (updates.description !== undefined) payload.description = updates.description;
    if (updates.tags !== undefined) payload.tags = updates.tags;

    const response = await this.client.request({
      method: 'PUT',
      url: `/detectors/${detectorId}`,
      data: payload
    });

    return response.data;
  }

  /**
   * Delete a detector.
   * 
   * @param detectorId Detector ID
   * @returns Promise resolving to true if deleted successfully
   */
  async deleteDetector(detectorId: string): Promise<boolean> {
    try {
      await this.client.request({
        method: 'DELETE',
        url: `/detectors/${detectorId}`
      });
      return true;
    } catch (error: any) {
      if (error.statusCode === 404) {
        throw new ResourceNotFoundError('Detector', detectorId);
      }
      throw error;
    }
  }

  // Training Operations

  /**
   * Train a detector with provided data.
   * 
   * @param detectorId Detector ID to train
   * @param dataset Training dataset
   * @param jobName Optional job name
   * @returns Promise resolving to training job information
   */
  async trainDetector(
    detectorId: string,
    dataset: Dataset,
    jobName?: string
  ): Promise<TrainingJob> {
    const payload = {
      detector_id: detectorId,
      dataset: this.formatDataset(dataset),
      ...(jobName && { job_name: jobName })
    };

    const response = await this.client.request({
      method: 'POST',
      url: '/training/jobs',
      data: payload
    });

    return this.formatTrainingJob(response.data);
  }

  /**
   * Get training job status and results.
   * 
   * @param jobId Training job ID
   * @returns Promise resolving to training job information
   */
  async getTrainingJob(jobId: string): Promise<TrainingJob> {
    try {
      const response = await this.client.request({
        method: 'GET',
        url: `/training/jobs/${jobId}`
      });
      return this.formatTrainingJob(response.data);
    } catch (error: any) {
      if (error.statusCode === 404) {
        throw new ResourceNotFoundError('Training Job', jobId);
      }
      throw error;
    }
  }

  /**
   * List training jobs with optional filtering.
   * 
   * @param options Filter and pagination options
   * @returns Promise resolving to paginated list of training jobs
   */
  async listTrainingJobs(options: {
    detectorId?: string;
    status?: string;
    page?: number;
    pageSize?: number;
  } = {}): Promise<PaginatedResponse<TrainingJob>> {
    const params: Record<string, any> = {
      page: options.page || 1,
      page_size: options.pageSize || 20
    };

    if (options.detectorId) params.detector_id = options.detectorId;
    if (options.status) params.status = options.status;

    const response = await this.client.request({
      method: 'GET',
      url: '/training/jobs',
      params
    });

    return {
      ...response.data,
      items: response.data.items.map((item: any) => this.formatTrainingJob(item))
    };
  }

  // Detection Operations

  /**
   * Detect anomalies in data using a trained detector.
   * 
   * @param detectorId ID of trained detector
   * @param dataset Data to analyze
   * @param options Detection options
   * @returns Promise resolving to detection results
   */
  async detectAnomalies(
    detectorId: string,
    dataset: Dataset,
    options: {
      returnScores?: boolean;
      threshold?: number;
    } = {}
  ): Promise<DetectionResult> {
    const payload = {
      detector_id: detectorId,
      dataset: this.formatDataset(dataset),
      return_scores: options.returnScores !== false,
      ...(options.threshold !== undefined && { threshold: options.threshold })
    };

    try {
      const response = await this.client.request({
        method: 'POST',
        url: '/detection/predict',
        data: payload
      });
      return this.formatDetectionResult(response.data);
    } catch (error: any) {
      if (error.statusCode === 404) {
        throw new ResourceNotFoundError('Detector', detectorId);
      }
      throw error;
    }
  }

  /**
   * Submit batch detection job for multiple datasets.
   * 
   * @param detectorId ID of trained detector
   * @param datasets List of datasets to process
   * @param jobName Optional job name
   * @returns Promise resolving to batch job information
   */
  async batchDetect(
    detectorId: string,
    datasets: Dataset[],
    jobName?: string
  ): Promise<any> {
    const payload = {
      detector_id: detectorId,
      datasets: datasets.map(dataset => this.formatDataset(dataset)),
      ...(jobName && { job_name: jobName })
    };

    const response = await this.client.request({
      method: 'POST',
      url: '/detection/batch',
      data: payload
    });

    return response.data;
  }

  // Experiment Management

  /**
   * Create and run a new experiment.
   * 
   * @param config Experiment configuration
   * @param dataset Dataset for the experiment
   * @returns Promise resolving to experiment information
   */
  async createExperiment(
    config: ExperimentConfig,
    dataset: Dataset
  ): Promise<any> {
    const payload = {
      name: config.name,
      description: config.description,
      algorithm_configs: config.algorithmConfigs.map(conf => ({
        algorithm_name: conf.algorithmName,
        hyperparameters: conf.hyperparameters || {},
        contamination_rate: conf.contaminationRate || 0.1,
        random_state: conf.randomState,
        n_jobs: conf.nJobs || 1
      })),
      evaluation_metrics: config.evaluationMetrics || ['auc_roc', 'precision', 'recall'],
      cross_validation_folds: config.crossValidationFolds || 5,
      random_state: config.randomState,
      parallel_jobs: config.parallelJobs || 1,
      optimization_enabled: config.optimizationEnabled || false,
      optimization_trials: config.optimizationTrials || 100,
      optimization_timeout: config.optimizationTimeout,
      dataset: this.formatDataset(dataset)
    };

    const response = await this.client.request({
      method: 'POST',
      url: '/experiments',
      data: payload
    });

    return response.data;
  }

  /**
   * Get experiment details and results.
   * 
   * @param experimentId Experiment ID
   * @returns Promise resolving to experiment information
   */
  async getExperiment(experimentId: string): Promise<any> {
    try {
      const response = await this.client.request({
        method: 'GET',
        url: `/experiments/${experimentId}`
      });
      return response.data;
    } catch (error: any) {
      if (error.statusCode === 404) {
        throw new ResourceNotFoundError('Experiment', experimentId);
      }
      throw error;
    }
  }

  /**
   * List experiments with optional filtering.
   * 
   * @param options Filter and pagination options
   * @returns Promise resolving to paginated list of experiments
   */
  async listExperiments(options: {
    page?: number;
    pageSize?: number;
    status?: string;
  } = {}): Promise<PaginatedResponse> {
    const params: Record<string, any> = {
      page: options.page || 1,
      page_size: options.pageSize || 20
    };

    if (options.status) params.status = options.status;

    const response = await this.client.request({
      method: 'GET',
      url: '/experiments',
      params
    });

    return response.data;
  }

  // Utility Methods

  /**
   * Validate dataset format and compatibility.
   * 
   * @param dataset Dataset to validate
   * @param algorithmName Optional algorithm to check compatibility
   * @returns Promise resolving to validation results
   */
  async validateDataset(
    dataset: Dataset,
    algorithmName?: string
  ): Promise<any> {
    const payload = {
      dataset: this.formatDataset(dataset),
      ...(algorithmName && { algorithm_name: algorithmName })
    };

    const response = await this.client.request({
      method: 'POST',
      url: '/validation/dataset',
      data: payload
    });

    return response.data;
  }

  /**
   * Get list of supported algorithms and their configurations.
   * 
   * @returns Promise resolving to list of algorithm information
   */
  async getSupportedAlgorithms(): Promise<any[]> {
    const response = await this.client.request({
      method: 'GET',
      url: '/algorithms'
    });

    return response.data.algorithms || [];
  }

  /**
   * Get detailed information about a specific algorithm.
   * 
   * @param algorithmName Name of the algorithm
   * @returns Promise resolving to algorithm details and parameter information
   */
  async getAlgorithmInfo(algorithmName: string): Promise<any> {
    try {
      const response = await this.client.request({
        method: 'GET',
        url: `/algorithms/${algorithmName}`
      });
      return response.data;
    } catch (error: any) {
      if (error.statusCode === 404) {
        throw new ResourceNotFoundError('Algorithm', algorithmName);
      }
      throw error;
    }
  }

  // Private helper methods

  /**
   * Format dataset for API consumption.
   */
  private formatDataset(dataset: Dataset): any {
    return {
      name: dataset.name,
      data: dataset.data,
      metadata: dataset.metadata || {},
      feature_names: dataset.featureNames,
      target_column: dataset.targetColumn
    };
  }

  /**
   * Format API training job response to TrainingJob.
   */
  private formatTrainingJob(data: any): TrainingJob {
    return {
      jobId: data.job_id,
      name: data.name,
      status: data.status,
      detectorConfig: {
        algorithmName: data.detector_config.algorithm_name,
        hyperparameters: data.detector_config.hyperparameters,
        contaminationRate: data.detector_config.contamination_rate,
        randomState: data.detector_config.random_state,
        nJobs: data.detector_config.n_jobs
      },
      datasetName: data.dataset_name,
      createdAt: data.created_at,
      startedAt: data.started_at,
      completedAt: data.completed_at,
      metrics: data.metrics ? this.formatModelMetrics(data.metrics) : undefined,
      modelPath: data.model_path,
      logs: data.logs || [],
      errorMessage: data.error_message,
      executionTime: data.execution_time,
      memoryUsage: data.memory_usage,
      cpuUsage: data.cpu_usage
    };
  }

  /**
   * Format API metrics response to ModelMetrics.
   */
  private formatModelMetrics(data: any): ModelMetrics {
    return {
      accuracy: data.accuracy,
      precision: data.precision,
      recall: data.recall,
      f1Score: data.f1_score,
      aucRoc: data.auc_roc,
      aucPr: data.auc_pr,
      contaminationRate: data.contamination_rate,
      anomalyThreshold: data.anomaly_threshold,
      customMetrics: data.custom_metrics || {}
    };
  }

  /**
   * Format API detection response to DetectionResult.
   */
  private formatDetectionResult(data: any): DetectionResult {
    return {
      anomalyScores: data.anomaly_scores,
      anomalyLabels: data.anomaly_labels,
      nAnomalies: data.n_anomalies,
      nSamples: data.n_samples,
      contaminationRate: data.contamination_rate,
      threshold: data.threshold,
      executionTime: data.execution_time,
      metadata: data.metadata || {}
    };
  }
}